pdns/pdns_recursor.cc

```diff
@@ -2897,50 +2929,94 @@ static void handleRunningTCPQuestion(int fd, FDMultiplexer::funcparam_t& var)
+            g_stats.cumulativeAnswers(spentUsec);
+	    dc->d_eventTrace.add(RecEventTrace::AnswerSent);
+
+	    if (t_protobufServers && dc->d_logResponse && !(luaconfsLocal->protobufExportConfig.taggedOnly && pbData && !pbData->d_tagged)) {
+	      struct timeval tv{0, 0};
+	      protobufLogResponse(dh, luaconfsLocal, pbData, tv, true, dc->d_source, dc->d_destination, dc->d_ednssubnet, dc->d_uuid, dc->d_requestorId, dc->d_deviceId, dc->d_deviceName, dc->d_meta, dc->d_eventTrace);
+	    }
+
+	    if (dc->d_eventTrace.enabled() && SyncRes::s_event_trace_enabled & SyncRes::event_trace_to_log) {
+	      g_log << Logger::Info << dc->d_eventTrace.toString() << endl;
+	    }
+          } // cache hit
+        } // query opcode
+
+        if(dc->d_mdp.d_header.opcode == Opcode::Notify) {
           if (!g_quiet) {
-            g_log<<Logger::Notice<<t_id<< " TCP question answered from packet cache tag="<<dc->d_tag<<" from "<<dc->d_source.toStringWithPort()<<(dc->d_source != dc->d_remote ? " (via "+dc->d_remote.toStringWithPort()+")" : "")<<endl;
+            g_log<<Logger::Notice<<t_id<< " got NOTIFY for "<<qname.toLogString()<<" from "<<dc->d_source.toStringWithPort()<<(dc->d_source != dc->d_remote ? " (via "+dc->d_remote.toStringWithPort()+")" : "")<<endl;
           }
 
-          bool hadError = sendResponseOverTCP(dc, response);
-          finishTCPReply(dc, hadError, false);
-          struct timeval now;
-          Utility::gettimeofday(&now, nullptr);
-          uint64_t spentUsec = uSec(now - start);
-          g_stats.cumulativeAnswers(spentUsec);
-          dc->d_eventTrace.add(RecEventTrace::AnswerSent);
-
-          if (t_protobufServers && dc->d_logResponse && !(luaconfsLocal->protobufExportConfig.taggedOnly && pbData && !pbData->d_tagged)) {
-            struct timeval tv{0, 0};
-            protobufLogResponse(dh, luaconfsLocal, pbData, tv, true, dc->d_source, dc->d_destination, dc->d_ednssubnet, dc->d_uuid, dc->d_requestorId, dc->d_deviceId, dc->d_deviceName, dc->d_meta, dc->d_eventTrace);
-          }
+          // send a message to the handler thread asking it
+          // to wipe all of the caches
+          ThreadMSG* tmsg = new ThreadMSG();
+          tmsg->func = [=]{ return pleaseWipeCaches(qname, true, 0xffff); };
+          tmsg->wantAnswer = false;
+          if(write(s_threadInfos[0].pipes.writeToThread, &tmsg, sizeof(tmsg)) != sizeof(tmsg)) {
+            delete tmsg;
 
-          if (dc->d_eventTrace.enabled() && SyncRes::s_event_trace_enabled & SyncRes::event_trace_to_log) {
-            g_log << Logger::Info << dc->d_eventTrace.toString() << endl;
+            unixDie("write to thread pipe returned wrong size or error");
           }
+
+          // the operation will now be treated as a Query, generating
+          // a normal response, as the rest of the code does not
+          // check dh->opcode, but we need to ensure that the response
+          // to this request does not get put into the packet cache
```

Comment:
Not a huge fan of the implication this comment makes. We might need to rethink the code structure (in a different PR)